### PR TITLE
CIRC-925 Improve handling of missing object reference for notice

### PR DESCRIPTION
**Pull request by Illia-Dolzhenko**
## Purpose
Improve handling of missing object reference for notice

## Approach
- Fix NPEs
- Add logging

Resolves: [CIRC-925](https://issues.folio.org/browse/CIRC-925)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 1]
```diff
@@ -78,10 +88,10 @@ public PatronActionSessionService(
   }
 
   public CompletableFuture<Result<Void>> endSession(String patronId,
-    PatronActionType actionType) {
+                                                    PatronActionType actionType) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:91]: Indentation

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 2]
```diff
@@ -66,6 +70,12 @@ public PatronActionSessionService(
   }
 
   public CompletableFuture<Result<LoanAndRelatedRecords>> saveCheckOutSessionRecord(LoanAndRelatedRecords records) {
+
+    if(records.getLoan() == null){
+      log.info("CheckOutSessionRecord is not saved, record doesn't have a valid loan.");
+      return completedFuture(of(() -> records));
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:76]: ofAsync(() -> records) ?

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 3]
```diff
@@ -66,6 +70,12 @@ public PatronActionSessionService(
   }
 
   public CompletableFuture<Result<LoanAndRelatedRecords>> saveCheckOutSessionRecord(LoanAndRelatedRecords records) {
+
+    if(records.getLoan() == null){
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:74]: Blank line, spaces

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 4]
```diff
@@ -124,6 +134,11 @@ public PatronActionSessionService(
 
     PatronSessionRecord recordSample = sessionRecords.get(0);
 
+    if (recordSample.getLoan() == null || recordSample.getLoan().getUser() == null) {
+      log.info("Notice was not sent. Session: {} doesn't have a valid loan or user.", recordSample.getId());
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:138]: Log what exactly is missing

File: src/main/java/org/folio/circulation/infrastructure/storage/sessions/PatronActionSessionRepository.java [hunk 5]
```diff
@@ -188,8 +188,14 @@ private boolean isPatronActionTypeSpecified(PatronActionType actionType) {
   private MultipleRecords<PatronSessionRecord> setUserForLoans(
     MultipleRecords<PatronSessionRecord> records, User user) {
 
-    return records.mapRecords(sessionRecord ->
-      sessionRecord.withLoan(sessionRecord.getLoan().withUser(user)));
+    return records.mapRecords(sessionRecord -> {
+        if (sessionRecord.getLoan() != null && sessionRecord.getPatronId() != null) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/infrastructure/storage/sessions/PatronActionSessionRepository.java:192]: Do we need sessionRecord.getPatronId() != null ?

File: src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java [hunk 6]
```diff
@@ -231,7 +231,7 @@ private void sendAvailableNotice(CheckInContext context, Request firstRequest) {
 
   private Result<CheckInContext> sendAvailableNotice(Request request, User user, CheckInContext context) {
     Item item = context.getItem();
-    if (item.isAwaitingPickup() && item.hasChanged()) {
+    if (item != null && item.isAwaitingPickup() && item.hasChanged()) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/resources/CheckInProcessAdapter.java:234]: Should we log an error when item != null ?

File: src/main/java/org/folio/circulation/resources/RequestNoticeSender.java [hunk 7]
```diff
@@ -69,7 +69,9 @@ public RequestNoticeSender(PatronNoticeService patronNoticeService,
       .withEventType(eventType)
       .withNoticeContext(createRequestNoticeContext(request))
       .build();
-    patronNoticeService.acceptNoticeEvent(requestCreatedEvent, NoticeLogContext.from(request));
+    if (item != null && requester != null) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:72]: Should we move this check to acceptNoticeEvent() ?

File: src/main/java/org/folio/circulation/resources/RequestNoticeSender.java [hunk 8]
```diff
@@ -105,13 +107,15 @@ public RequestNoticeSender(PatronNoticeService patronNoticeService,
...
     User requester = request.getRequester();
 
-    PatronNoticeEvent requestCancelledEvent = new PatronNoticeEventBuilder()
-      .withItem(item)
-      .withUser(requester)
-      .withEventType(NoticeEventType.REQUEST_CANCELLATION)
-      .withNoticeContext(createRequestNoticeContext(request))
-      .build();
-    patronNoticeService.acceptNoticeEvent(requestCancelledEvent, NoticeLogContext.from(request));
+    if(requester != null && item != null) {
```
Comment (alexanderkurash) [inline src/main/java/org/folio/circulation/resources/RequestNoticeSender.java:110]: Same question here + space and order of checks

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 9]
```diff
@@ -107,9 +113,12 @@ public PatronActionSessionService(
     MultipleRecords<PatronSessionRecord> records) {
 
     if (records == null) {
+      log.info("Records are empty, nothing to delete.");
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:116]: Just by looking at this phrase in the log, it's not that easy to understand what exactly "Records" is. Should we maybe add so me more details?

File: src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java [hunk 10]
```diff
@@ -118,12 +127,28 @@ public PatronActionSessionService(
     MultipleRecords<PatronSessionRecord> records) {
 
     if (records.isEmpty()) {
+      log.info("Records are empty, notices will not be sent.");
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/notice/session/PatronActionSessionService.java:130]: Just by looking at this phrase in the log, it's not that easy to understand what exactly "Records" is. Should we maybe add so me more details?

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**